### PR TITLE
[ADD] models: add methods to efficiently concatenate and union recordsets

### PR DESCRIPTION
**Pull request by rco-odoo**
As the title says:

``` python
recs0.concat(recs1, ..., recsn)     # same as recs0 + recs1 + ... + recsn
recs0.union(recs1, ..., recsn)      # same as recs0 | recs1 | ... | recsn
```

### CODE REVIEW

**Comment by rco-odoo** [comment]
@odony @KangOl ping

File: odoo/models.py [hunk 1]
```diff
@@ -5105,27 +5103,30 @@ def __add__(self, other):
...
     def __and__(self, other):
         """ Return the intersection of two recordsets.
-            Note that recordset order is not preserved.
+            Note that first occurrence order is preserved.
         """
         if not isinstance(other, BaseModel) or self._name != other._name:
             raise TypeError("Mixing apples and oranges: %s & %s" % (self, other))
-        return self.browse(OrderedSet(self._ids) & OrderedSet(other._ids))
+        other_ids = set(other._ids)
+        return self.browse(OrderedSet([id for id in self._ids if id in other_ids]))
```
Comment (KangOl) [inline odoo/models.py]: use a generator instead of a list comprehension.

**Comment by odony** [comment]
Looks good! I guess the implementation of `__or__` and `__and__` in `OrderedSet` is not strictly necessary anymore, although it does not hurt.
By the way, maybe `BaseModel.__add__` and `BaseModel.__or__` could be implemented with  `self.concat(other)` and `self.union(other)`, respectively, to avoid 2 similar implementations for the same thing?

**Comment by rco-odoo** [comment]
@odony @KangOl I refactored a bit the implementation:
* recordset `a + b` is implemented as `a.concat(b)`
* recorset `a | b` is implemented as `a.union(b)`
* both `concat()` and `union()` check their arguments types

I also improved the implementation of `OrderedSet` such that it implements the full `set` API.

**Comment by dreispt** [comment]
Missing docs 😢

**Pull request merged**